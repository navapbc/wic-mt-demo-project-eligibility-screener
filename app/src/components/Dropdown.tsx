import { Trans } from 'next-i18next'
import { ChangeEvent, ReactElement } from 'react'

import Required from '@components/Required'

interface Props<T> {
  handleChange: (
    e: ChangeEvent<HTMLSelectElement> & { target: { value: T } }
  ) => void
  id: string
  labelKey: string
  options: string[]
  required?: boolean
}

// This component expects pre-translated option strings.
// @TODO: This should be refactored if its ever used with non-integer options.
const Dropdown = <T extends string>(props: Props<T>): ReactElement => {
  const { handleChange, id, labelKey, options, required } = props

  return (
    <>
      <label className="usa-label" htmlFor={id}>
        <Trans i18nKey={labelKey} />
        {required && <Required />}
      </label>
      <select className="usa-select" id={id} onChange={handleChange}>
        <option value={undefined}>
          -&nbsp;
          <Trans i18nKey="select" />
          &nbsp;-
        </option>
        {options.map((option: string) => (
          <option value={option} key={option}>
            {option}
          </option>
        ))}
      </select>
    </>
  )
}

export default Dropdown
