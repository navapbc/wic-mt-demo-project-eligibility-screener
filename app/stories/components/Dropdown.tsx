import { ComponentMeta, ComponentStory } from '@storybook/react'

import DropdownComponent from '@components/Dropdown'

export default {
  title: 'Components',
  component: DropdownComponent,
} as ComponentMeta<typeof DropdownComponent>

const Template: ComponentStory<typeof DropdownComponent> = (args) => (
  <DropdownComponent {...args} />
)

export const Dropdown = Template.bind({})

Dropdown.args = {
  id: 'test',
  label: 'Select',
  options: ['1', '2', '3'],
}
