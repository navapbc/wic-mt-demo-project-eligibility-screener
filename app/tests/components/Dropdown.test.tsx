import { render, screen } from '@testing-library/react'

import Dropdown, { DropdownProps } from '@components/Dropdown'

import { testSnapshot } from '../helpers/sharedTests'

const testProps: DropdownProps = {
  handleChange: (e) => {},
  id: 'test-id',
  labelKey: 'dropdown label',
  options: ['a', 'b', 'c'],
}

it('should match snapshot', () => {
  testSnapshot(<Dropdown {...testProps} />)
})

it('should match display required marker if required is true', () => {
  render(<Dropdown {...testProps} required={true} />)
  const required = screen.getByText('*')
  expect(required).toBeInTheDocument()
})

it('should show the selected element if passed in', () => {
  render(<Dropdown {...testProps} selectedOption="c" />)
  const required = screen.getByRole('option', {
    name: 'c',
  }) as HTMLOptionElement
  expect(required.selected).toBe(true)
})
