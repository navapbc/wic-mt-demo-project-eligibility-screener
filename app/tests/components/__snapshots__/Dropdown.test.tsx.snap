// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <label
    class="usa-label"
    for="test-id"
  >
    dropdown label
  </label>
  <select
    class="usa-select"
    id="test-id"
    name="test-id"
  >
    <option
      value=""
    >
      - 
      Select
       -
    </option>
    <option
      value="a"
    >
      a
    </option>
    <option
      value="b"
    >
      b
    </option>
    <option
      value="c"
    >
      c
    </option>
  </select>
</div>
`;
